common/flatpak-run.c

```diff
@@ -3686,7 +3712,23 @@ flatpak_run_app (FlatpakDecomposed *app_ref,
   if (extra_context)
     flatpak_context_merge (app_context, extra_context);
 
-  runtime_files = flatpak_deploy_get_files (runtime_deploy);
+  original_runtime_files = flatpak_deploy_get_files (runtime_deploy);
+
+  if (custom_runtime_usr != NULL)
+    {
+      runtime_files = g_file_new_for_path (custom_runtime_usr);
+      /* For now we assume that if the caller wanted to update the
+       * ld.so.cache, it would already have done so; that's certainly
+       * true for pressure-vessel. */
+      use_ld_so_cache = FALSE;
```

Comment:
This will also remove the use of the ld.so cache for the app. But I guess its not really possible to run in some "mixed mode".

Reply:
Yeah, I need to think about this some more.

Disabling the ld.so cache regeneration and relying on `LD_LIBRARY_PATH` to find the imported graphics drivers is definitely the conservative, learn-to-walk-before-we-run option, and is how pressure-vessel works right now.

I'm trying to teach pressure-vessel to generate its own `ld.so.cache` in its specially-prepared `/usr` when it launches games outside Flatpak, and I had thought that I'd be able to do that in pressure-vessel in the Flatpak case too, and hand it over to Flatpak already prepared - but now that I think about it more, I might not be able to do that without entering a container. If I can't, and I want to use a `ld.so.cache`, I'll have to either teach Flatpak to do that (in a way that is appropriate for the `/usr` I'm giving it), or put `ld.so.cache` somewhere that is modifiable by the subsandbox (to which the one in `/etc` would be a symlink) and tell the launched command to regenerate it before running the actual game.

The code in pressure-vessel that generates its own `ld.so.cache` also doesn't work yet (graphics drivers don't load successfully, and I haven't been able to pin down why).

Reply:
I looked at enabling the ld.so cache generation, but the logic for cache invalidation assumes that the `ld.so.cache` is based on the app deployment, the app extensions, the runtime deployment and the runtime extensions. To be able to do the cache invalidation correctly, the caller would have to be able to pass in unique identifiers for its `/app` and `/usr` that would be hashed into the `ld.so.cache` checksum.

However, the Steam container runtime is dynamically set up based on not just a runtime, but also the graphics driver stack that it's borrowing, which might come from the Flatpak runtime, the host system or some third location - so there's no obvious unique ID to give it, and in practice I'd just be generating a UUID, which would result in the ld.so cache being invalidated and regenerated every time. The Steam runtime is also not stored in a content-addressed way, so we can't really rely on it not being modified locally.

Do we want to go to those lengths?

Reply:
The ld.so cache generation code also makes some assumptions about the ld.so.conf and ld.so.cache setup, such as "it's appropriate to enable all the extensions" and "we'll use `/run/flatpak/ld.so.conf.d/runtime-*.conf`" - although those are things I could work around in the runtime, if I need to.